File: src/SmapiControl.java
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.google.gson.stream.JsonReader;

import java.io.*;
import java.lang.reflect.Type;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

public class SmapiControl {
    // TODO: implement checkSmapiInstall
    private static void checkSmapiInstall() {

    }

    // TODO: implement checkSmapiVersion
    private static void checkSmapiVersion() {

    }

    // TODO: implement downloadSmapi
    // TODO: Deal with github assets. This currently downloads source
    private static Path downloadSmapi() {
        SerializedTypes.SmapiReleaseResponse response = getReleaseResponse();
        String assetsUrl = response.assets_url;
        String tag = response.tag_name;
        String zipUrl = null;
        Path assetsPath = null;
        JsonReader jsonReader = null;
        Path result = null;
        final String assetsFilename = "immAssets.json";
        final String zipFilename = "immSmapi.zip";

        try {
            assetsPath = sendGetAndSave(assetsUrl, assetsFilename);
        } catch (IOException e) {
            e.printStackTrace();
        }

        try {
            assert assetsPath != null;
            jsonReader = new JsonReader(new FileReader(assetsPath.toFile()));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        assert jsonReader != null;
        Type listType = new TypeToken<ArrayList<SerializedTypes.SmapiAssetResponse>>() {
        }.getType();
        List<SerializedTypes.SmapiAssetResponse> assetList = new Gson().fromJson(jsonReader, listType);
        for (SerializedTypes.SmapiAssetResponse element : assetList) {
            if (element.name.equals("SMAPI-" + tag + ".zip")) {
                zipUrl = element.browser_download_url;
            }
        }

        if (zipUrl == null || zipUrl.isEmpty()) {
            // TODO: Proper error handling
            System.err.println("Couldn't find SMAPI download package.");
            System.exit(-1);
        } else {
            try {
                result = sendGetAndSave(zipUrl, zipFilename);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }

        return result;

    }

    private static SerializedTypes.SmapiReleaseResponse getReleaseResponse() {
        final String url = "https://api.github.com/repos/Pathoschild/SMAPI/releases/latest";
        final String filename = "immSmapiResp.json";
        Path responsePath = null;
        JsonReader jsonReader = null;

        try {
            responsePath = sendGetAndSave(url, filename);
        } catch (IOException e) {
            e.printStackTrace();
        }

        try {
            assert responsePath != null;
            jsonReader = new JsonReader(new FileReader(responsePath.toFile()));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        assert jsonReader != null;
        return new Gson().fromJson(jsonReader, SerializedTypes.SmapiReleaseResponse.class);
    }

    private static Path sendGetAndSave(String url, String filename) throws IOException {
        URL urlObj;
        HttpURLConnection connection;
        Path savePath;
        InputStream in;
        FileOutputStream out;
        int responseCode;
        int read;
        byte[] buffer;

        urlObj = new URL(url);
        connection = (HttpURLConnection) urlObj.openConnection();

        connection.setRequestMethod("GET");
        connection.setRequestProperty("User-Agent", "Iridium Mod Manager");

        responseCode = connection.getResponseCode();
        if (responseCode != 200) {
            // TODO: Handle bad response
            System.err.println("Bad server response. Response: " + responseCode + " from address " + url);
            System.exit(-1);
        }
        savePath = Paths.get(System.getProperty("java.io.tmpdir"), filename);
        if (savePath.toFile().exists()) {
            savePath.toFile().delete();
        }

        in = connection.getInputStream();
        out = new FileOutputStream(savePath.toFile());

        read = 0;
        buffer = new byte[32768];
        while ((read = in.read(buffer)) > 0) {
            out.write(buffer, 0, read);
        }

        out.close();
        in.close();
        return savePath;
    }

    private static void installSmapi() {
        // TODO: unzip SMAPI
        // TODO: copy all files; Mono if *nix, Windows if win
        // TODO: rename .exe if *nix
        // TODO: show "edit your launch settings" if windows
    }
}

File: src/SmapiControl.cpp
//
// Created by Colten Sullivent on 10/5/17.
//

#include "../include/SmapiControl.h"

void SmapiControl::checkSmapiInstall() {

}

void SmapiControl::checkSmapiVersion() {
// tag_name
}

boost::filesystem::path SmapiControl::downloadSmapi() {
  SerializedTypes::SmapiReleaseResponse::obj response = getReleaseResponse();
  std::string url = "https://github.com/Pathoschild/SMAPI/releases/download/";
  url += response.tag_name;
  url += "/SMAPI-";
  url += response.tag_name;
  url += ".zip";

  return sendGetAndSave(url, "smapi.zip");
}

void SmapiControl::installSmapi() {
  boost::filesystem::path zipPath = SmapiControl::downloadSmapi();

  zipper::Unzipper unzipper(zipPath.string());
  unzipper.extract(boost::filesystem::temp_directory_path().string() + "/smapi");
  unzipper.close();
}

SerializedTypes::SmapiReleaseResponse::obj SmapiControl::getReleaseResponse() {
  const std::string url = "https://api.github.com/repos/Pathoschild/SMAPI/releases/latest";
  const std::string filename = "immSmapiReponse.json";

  boost::filesystem::path path = sendGetAndSave(url, filename);
  SerializedTypes::SmapiReleaseResponse::obj response;
  boost::filesystem::ifstream i(path);
  json j;
  i >> j;
  i.close();
  response = j;

  return response;
}

boost::filesystem::path SmapiControl::sendGetAndSave(std::string url, std::string filename) {
  CURL *curl;
  CURLcode res;
  FILE *fp;
  std::string fullpath = boost::filesystem::temp_directory_path().string() + filename;

  curl = curl_easy_init();
  if (curl) {
    fp = fopen(fullpath.c_str(), "wb");
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_USERAGENT, "Iridium Mod Manager");
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
    res = curl_easy_perform(curl);
    if (res != CURLE_OK) {
      // TODO: Proper error handling
      fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
    }
    curl_easy_cleanup(curl);
    fclose(fp);
  }

  return boost::filesystem::path(fullpath);
}
